usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -283,43 +288,82 @@ def populate_response(self, results: list, has_next: bool) -> dict:
+            d = datetime.strptime(date, "%Y-%m-%d")
+            date = int(d.timestamp() * 1000)
+        return date
+
     def query_elasticsearch(self) -> list:
         filter_query = QueryWithFilters.generate_awards_elasticsearch_query(self.filters)
         sort_field = self.get_elastic_sort_by_fields()
         sorts = [{field: self.pagination["sort_order"]} for field in sort_field]
-        search = (
-            (
+        record_num = (self.pagination["page"] - 1) * self.pagination["limit"]
+
+        if record_num >= settings.ES_AWARDS_MAX_RESULT_WINDOW and (
+            self.last_record_unique_id is None and self.last_record_sort_value is None
+        ):
+            logger.warning(
+                "WARNING: Jumping to page {page} with page size {limit}. First record number: {record}. Retrieving records past {es_limit} records will have a performance hit when using Elasticsearch".format(
+                    page=self.pagination["page"],
+                    limit=self.pagination["limit"],
+                    record=self.pagination["lower_bound"],
+                    es_limit=settings.ES_AWARDS_MAX_RESULT_WINDOW,
+                )
+            )
+            sort_by_fields = self.get_sort_by_fields()
+            sort_by_fields.append("award_id")
+            database_fields = self.get_database_fields()
+            base_queryset = self.constants["filter_queryset_func"](self.filters)
+            queryset = self.annotate_queryset(base_queryset)
+            queryset = self.custom_queryset_order_by(queryset, sort_by_fields, self.pagination["sort_order"])
+            queryset = queryset.values(*list(database_fields))[record_num - 1 : record_num]
+
+            if len(queryset) != 1:
+                return {}
+            results = [
+                self.date_to_epoch_millis(queryset[0].get(self.get_sort_by_fields()[0])),
+                queryset[0].get("award_id"),
+            ]
+            search = (
                 AwardSearch()
                 .filter(filter_query)
                 .sort(*sorts)
-                .extra(search_after=[self.last_value, self.last_id])[0 : self.pagination["limit"]]
+                .extra(search_after=[*results])[: self.pagination["limit"] + 1]
             )
-            if self.last_value and self.last_id
-            else (
+            response = search.handle_execute()
+            return response
+        search = (
+            (
                 AwardSearch()
                 .filter(filter_query)
-                .sort(*sorts)[
-                    ((self.pagination["page"] - 1) * self.pagination["limit"]) : (
-                        ((self.pagination["page"] - 1) * self.pagination["limit"]) + self.pagination["limit"]
-                    )
-                ]
+                .sort(*sorts)
+                .extra(
+                    search_after=[self.date_to_epoch_millis(self.last_record_sort_value), self.last_record_unique_id]
```

Comment:
So here, in the scenario where the API caller has _provided_ `last_record_sort_value` in the API request POST json...

Are we expecting this value to be provided as a normal date? Or as epoch-millis?

Because, we _return_ (in the HTTP Response) it from this endpoint as epoch-millis ... so for consistency, I would expect them to provide it back to us in the same format (in which case it does not need to be converted).

Reply:
Remove the conversion on the API-provided value, kept it for the postgres lookup, which does need to be converted from date to epoch-millis